### DISCUSSION TITLE
Feature Request: Pass Through Additional Metadata (`id`, `chat_id`, `session_id`) to Custom Backend

### DISCUSSION THREAD

**Discussion by GregorBiswanger** [open-webui/open-webui#6999] [Ideas]
## Important Notes

As a .NET developer and enthusiastic user of Open WebUI, I’ve developed a NuGet package, [OllamaApiFacade](https://github.com/GregorBiswanger/OllamaApiFacade), that allows .NET developers to integrate Open WebUI as a ready-to-use frontend UI while implementing their own backend logic in C#. I also plan to create a similar package for Node.js, enabling developers to focus on building AI solutions without worrying about frontend or communication logic.

Open WebUI already provides excellent features for communicating with backends like Ollama or OpenAI. However, these backends are stateless, and adding the ability to pass metadata such as `id`, `chat_id`, and `session_id` through to custom backends would simplify state management and enhance the developer experience.

---

### **Is your feature request related to a problem? Please describe.**

Currently, when Open WebUI forwards requests to custom backends, critical metadata such as `id`, `chat_id`, and `session_id` is not included. For developers building stateful backend logic, the absence of this information makes it difficult to maintain user or session context.

As a fan of Open WebUI, I find it challenging to handle state management in my backend effectively without these fields. This makes the development process more complex and detracts from focusing on core AI logic.

---

### **Describe the solution you'd like**

I propose that Open WebUI includes an option or standard behavior to forward additional metadata (`id`, `chat_id`, `session_id`) when sending requests to custom backends. This would allow developers to:

1. Retain session and user context across requests.
2. Implement robust state management with minimal effort.
3. Focus on their own logic instead of building complex workarounds for tracking state.

This feature would make Open WebUI even more developer-friendly and powerful for integrating custom backends.

---

### **Describe alternatives you've considered**

- **Manual Injection of `session_id` or `chat_id`:** I have attempted to explore workarounds to manually inject these fields into requests or simulate them. However, without native support from Open WebUI, this approach is infeasible, as it is almost impossible to reliably identify and process requests.

- **Custom Pipelines:** While Open WebUI allows building custom pipelines, this approach feels overly complex for someone without Python experience. It would require significant effort to implement a pipeline just to pass these fields. A built-in solution would greatly simplify the process and make it accessible to a broader audience.

In summary, I have not found any viable alternative that would allow me to handle identifiable requests or manage state efficiently.

---

### **Additional context**

I am a big fan of Open WebUI and deeply impressed by its capabilities and flexibility. Adding support for metadata pass-through would align with its vision of empowering developers to create innovative AI solutions. This feature would make Open WebUI even more valuable for developers across different platforms, whether in .NET, Node.js, or others.

Thank you for considering this request! I would be happy to contribute to further discussions or testing this feature.

**Comment by tjbck** [comment]
You should be able to implement this function as a Pipe.

**Comment by GregorBiswanger** [comment]
Thank you for your response! I completely understand the flexibility and power of using Pipes in Open WebUI. However, the primary goal of this feature request is to lower the barrier for developers in ecosystems like .NET, Node.js, and others to seamlessly integrate Open WebUI into their development workflows.

The idea is to make Open WebUI a "ready-to-go" solution that developers can use out of the box without having to set up additional containers or write Python scripts to forward existing metadata such as session_id, chat_id, and id. These fields are often critical for backend state management in AI solutions, where stateless communication is a standard practice. By not passing these already-available fields directly, it creates a hurdle for developers to simply get started.

Personally, I’m not familiar with Python, so implementing a Pipe to forward this information adds a layer of complexity that prevents me from focusing on building my backend logic. It also adds friction for other developers who want to adopt Open WebUI as a plug-and-play solution for their environments.

By making this metadata pass-through a standard feature, Open WebUI can provide a seamless experience for developers, enabling them to concentrate on creating innovative AI solutions without additional setup or learning curves.

I’m more than happy to contribute to testing or brainstorming how this feature could be implemented in the most efficient way. Let me know how I can support this effort!

**Comment by DmitriyAlergant** [comment]
@tjbck not really. 

The Pipe function can receive many interesting parameters, but NOT the metadata dict (that has the chat_id and session_id). 
The metadata object is popped from request before it is passed to the pipe functions.

I raised the same concern recently: https://github.com/open-webui/open-webui/discussions/6666 and even raised a PR (https://github.com/open-webui/open-webui/pull/6665) which you closed without comments...

**Reply by tjbck** [reply]
Good catch, e3485d2d88e736a4e3fabc394579fd1ed9951ef7 should address this.

**Comment by DmitriyAlergant** [comment]
Thanks ❤️

**Comment by GregorBiswanger** [comment]
Just briefly for dummies. What does that mean? That it will be included in the next release, right? 😅

**Reply by DmitriyAlergant** [reply]
In the next release, he has added a capability for Pipe Functions to receive the Metadata dict (id, chat_id, session_id) that the Pipe function can then pass over to the provider, if you so desire. You would still need to use Pipe functions.

You don't need to develop from scratch, there are community-provided pipes for various providers including OpenAI itself. You just disable built-in integration with OpenAI, and install a Pipe instead. Like this one from the Community Hub: https://openwebui.com/f/hub/openai_manifold - very simple.    Then if you want to pass additional parameters you can easily add that.  With the next release, because currently Pipe functions cannot receive the metadata dict. After the maintainer's change, it becomes possible.

We are using a more sophisticated set of pipes (https://github.com/DmitriyAlergant-T1A/open-webui-cost-tracking-manifolds) that include usage and cost tracking. These are not published to the community hub.

**Reply by GregorBiswanger** [reply]
Sounds good! Anything that would be possible without additional effort or simple configuration from the outside of a Docker container would be gold - promoting Open WebUI as a simple "ready-to-go" variant.

So I'm happy that something is happening. Better than nothing. But activating simple metadata, e.g. via environment variables, would really be gold here.

I would really like to avoid creating an extra pipeline container. Certainly interesting for users who just want to bind to language models and control processes via Python.

But anyway, a **big thank you to everyone** involved in the project, **I'll definitely push you!** 🚀🚀🚀

**Reply by DmitriyAlergant** [reply]
You are still confusing "Pipelines" vs "Functions". See https://docs.openwebui.com/features/plugin/#functions

Functions are very easy to deploy to OpenWebUI by an admin via its Web UI, literally in a few clicks. It's not a  lot of complexity, not at all.

The standalone Pipelines module (running in another container), to the contrary, is heavy-lifting and introduces lots of complexity. I am not using it and you should not either.

**Reply by GregorBiswanger** [reply]
Ahh, so this is enabled via functions...  Yes then, I can hardly wait until the next release is published! 🥳🥳

As I said, I have a lot planned so that I can advertise your project properly. Because it really offers a lot of benefits and has been implemented really well. 🥰

**Reply by DmitriyAlergant** [reply]
You can begin with leveraging Functions now to get comfortable with the end-to-end solution and have it in place; 
And only add metadata pass-through later with the upcoming release. 

Remember, using pipe function to integrate with OpenAI means you would need to disable the built-in integration in Admin Settings otherwise you would see duplicate models in the catalog.  Also any model-level settings (System Prompts, Descriptions, etc) if you made any changes from default settings, you would need to re-apply again to the models provided by the function.

Also FYI, functions only work for Models Chat Completion.     The other avenues where Web UI optionally integrates with OpenAI (Audio STT or TTS, Images Generation, Embeddings, etc) remain functional if you had them enabled and work normally even if the main OpenAI integration (Admin Panel -> Connections) id disabled.

**Reply by GregorBiswanger** [reply]
I saw you released 4.0. I tried it with functions, but it didn't work.

To be honest, I'm trying it via ChatGPT, which automatically generates useful information for me, because I'm not really good at Python. Or is it not possible with 4.0 yet?

**Reply by GregorBiswanger** [reply]
I've been experimenting for hours and unfortunately haven't succeeded. 😪

**Reply by DmitriyAlergant** [reply]
I have not released anything, I am not the maintainer.

You haven't succeeded in what, having OpenAI Pipe function to work? 

You don't need to develop it, you need to get existing code from the community. 
Like this one: https://openwebui.com/f/hub/openai_proxy_pipe      Just click "Get' and deploy it, configure API key in the Valve.

It just works; Yes, on 0.4.0; What's there not to work?

**Reply by GregorBiswanger** [reply]
I’m currently running version 4.0 as a container and encountered the following issue:

- I generated the Function Pipe with the help of ChatGPT: [Code on Pastebin](https://pastebin.com/PQai5gY9).
- In the Admin menu under “Functions,” I created and activated the new function.
- I disabled the OpenAI integration to avoid duplicate calls, as you had mentioned earlier.

However, when writing in the chat, my backend still only receives a stripped-down JSON. It does not include the same information exchanged between the web app and the Open WebUI backend. The JSON looks like this:

```json
{
  "model": "dotnetapi:latest",
  "messages": [
    {
      "role": "user",
      "content": "hello"
    }
  ],
  "options": {},
  "stream": true
}
```

The essential fields like `id`, `session_id`, and `user_id` are missing, which are crucial for my processing.

Am I missing something, or is there a way to include the missing information in the backend requests?

Thanks in advance for your help!

**Comment by DmitriyAlergant** [comment]
In the recent version, your pipe should be able to receive the "metadata" dict containing chat_id, session_id etc. when you add it as a parameter to the function definition:

` def pipe(self, body: dict, __user__: dict, __metadata__: dict)`
- corrected
 
Note, i have not tested that myself but that is how it is supposed to work based on my discussion with the maintainer in this issue, and the committed change.

This is really where you need to start looking into the Open WebUI's source code to understand what it is doing behind the scenes to figure this out.  Also as I said above, it is better to begin with existing implementation of pipes rather then creating one from scratch.

**Reply by tjbck** [reply]
`__metadata__`* 😊

**Reply by GregorBiswanger** [reply]
I tried this. Unfortunately it doesn't work with the latest version of Open WebUI either. 😔

```
"""
title: AddMetadataWithIDs
author: yourname
author_url: https://github.com/yourgithub
version: 1.2.0
date: 2024-11-22
license: GPLv3
description: A Pipe that adds user, metadata, and specific fields (id, chat_id, session_id) to requests.
"""

from pydantic import BaseModel, Field
from typing import Union, Optional, Generator, Iterator
import json


class Pipe:
    class Valves(BaseModel):
        debug: bool = Field(
            default=False, description="True to enable debug logs."
        )

    def __init__(self):
        self.type = "manifold"  # Integration type
        self.id = "add_metadata_with_ids"  # Alphanumeric ID for the pipe
        self.name = "Add Metadata with IDs"  # Display name for the pipe
        self.valves = self.Valves()

    def pipes(self):
        """
        Define available pipes.
        """
        return [{"id": self.id, "name": self.name}]

    def pipe(
        self, body: dict, __user__: dict, metadata: dict
    ) -> Union[dict, Generator, Iterator]:
        """
        Add user information and metadata (including id, chat_id, session_id) to the body.
        """
        def log(message: str):
            if self.valves.debug:
                print(f"DEBUG: {message}")

        log(f"Original body: {body}")
        log(f"Metadata: {metadata}")
        log(f"User: {__user__}")

        # Add user information
        if __user__:
            body["user"] = __user__.get("name", "unknown_user")

        # Add specific fields from metadata
        body["id"] = metadata.get("id", body.get("id", "default_id"))
        body["chat_id"] = metadata.get("chat_id", body.get("chat_id", "default_chat_id"))
        body["session_id"] = metadata.get("session_id", body.get("session_id", "default_session_id"))

        # Add metadata directly to the body
        body["metadata"] = metadata

        log(f"Updated body: {body}")
        return body

```

**Comment by samling** [comment]
I'm interested in the solution to this as well. I have open-webui sending to litellm which sends traces to langfuse; I'd like to set tags and other data in langfuse for requests from OWUI, which I'm attempting to do through a filter. I'm doing essentially the same thing as @GregorBiswanger without any luck; looking at the debug logs in litellm, `__metadata__` nor `body["metadata"]` make it to litellm.

**Reply by DmitriyAlergant** [reply]
Is your connection to litellm via a Pipe Function?

It works for me.

1) Open WebUI 0.4.1

2) A Function defines a Pipe class, and the "pipe" function receives ```__metadata__``` dict as an input

3) The pipe can extract the chat ID from ```__metadata__``` and use it as it needs to. In my case, I'm using it for logging to database. And it works for me. I am seeing the Chat IDs logged to the database.  In other situations it could have been used to send to the API provider (ex: langfuse).

It works for me in a Pipe. It may not be working for a Filter though - I have not checked that.

My implementation of (multiple) Pipe functions that successfully receives and processes ```__metadata__``` is here: 
https://github.com/DmitriyAlergant-T1A/open-webui-cost-tracking-manifolds

**Reply by samling** [reply]
Nope, my connection is just by configuring it in the connections page, since it's a compatible openai api endpoint.

Very interesting, thanks for sharing. I'll take a look at your pipes and compare to what I'm doing. A condensed, very simple version is basically like this:

```
class Pipeline:
    class Valves(BaseModel):
        priority: int = 0
        pipelines: List[str] = []

    def __init__(self):
        # Initialize
        self.type = "filter"
        self.valves = self.Valves(
            **{
                "pipelines": ["*"]
              }

    async def inlet(
        self,
        body: dict,
        __user__: dict,
        __metadata__: dict,
    ) -> dict:
        __metadata__["tags"] = ["open-webui"]
```

I can verify that `__metadata__` itself gets updated, but none of the contents of `__metadata__` are included in the request to litellm, based on looking at the debug logs of what litellm receives. Litellm only receives `stream`, `model`, `messages` and, if I set it, `tags` (which doesn't get used by litellm or langfuse as-is). An example request as received by litellm looks like this:

```
{
    "stream": true,
    "model": "gpt-4o-2024-11-20",
    "messages": [
        {"role": "user", "content": "Testing"}
    ],
    "proxy_server_request": {
        "url": "http://litellm.litellm:4000/v1/chat/completions",
        "method": "POST",
        "headers": {
            "host": "litellm.litellm:4000",
            "content-type": "application/json",
            "accept": "*/*",
            "accept-encoding": "gzip, deflate",
            "user-agent": "Python/3.11 aiohttp/3.10.8",
            "content-length": "124"
        },
        "body": {
            "stream": true,
            "model": "gpt-4o-2024-11-20",
            "messages": [
                {"role": "user", "content": "Testing"}
            ]
        }
    }
}
```

**Reply by DmitriyAlergant** [reply]
There was no change or an update by the maintainer to include any additional metadata to "Standard" OpenAI connection. 

The only change discussed in this thread was making the ```__metadata__``` parameter available to custom Functions in Python.

Furthermore, some of the API providers may be specifically unhappy with any body elements they did not expect. AFAIK OpenAI is OK ignoring them, but e.g. Anthropic API is failing on anything it does not expect. So I don't think the maintainer should add **any** additional stuff (not specifically documented by OpenAI in their API specs) to built-in calls to OpenAI. At the very least, this may unexpectedly disclose information to a 3rd party.

You want more control and customizations over how Open WebUI integrates with providers? Stop using built-in integration, and use Pipe function(s) instead.

**Reply by samling** [reply]
Got it, I misunderstood the original ask of the thread. Thanks.

**Reply by GregorBiswanger** [reply]
@DmitriyAlergant-T1A Unfortunately, I haven't been able to find a solution to my problem yet. Can you tell from my code above why I can't pass on the metadata? Do you see the error? I need this urgently. So that I can work with my state manager in the backend. 😪

**Reply by samling** [reply]
@GregorBiswanger In your code it looks like you used `metadata` instead of `__metadata__`, does it not work with the latter?

**Reply by DmitriyAlergant** [reply]
@samling @GregorBiswanger 

Yeah that will fix it;  But his pipe is still essentially a stub, it will only be reflecting the body back to the chat; I hope that's intended and he has the plan of completing the pipe to actually call the LLM provider;  Or use one of the numerous existing implementations linked above in this thread.

**Reply by GregorBiswanger** [reply]
I'll try it out

**Reply by GregorBiswanger** [reply]
@samling @DmitriyAlergant-T1A I tried it: [https://pastebin.com/iF8LW2e4](https://pastebin.com/iF8LW2e4) 

In the container log you can see data with debug. But my backend still gets the following JSON without any further data:
`{"model": "dotnetapi:latest", "messages": [{"role": "user", "content": "Hello"}], "options": {}, "stream": true}`

**Reply by DmitriyAlergant** [reply]
Your backend cannot get anything from this function. This pipe function does nothing but reflect back the request body.

In your Open WebUI in the list of models if you find the model provided by this pipe ("Add Metadata with IDs") and talk to it, all you will see back in the assistant responses would be Updated body: {body} with some payload body.  No AI.

If your backend still gets ANYTHING this means you have still have the built-in OpenAI or OLLama connection enabled. Your chats still use these models (provided by built-in connections) directly and not using this PIpe.

The Manifold Pipe function does not augment the built-in integration with OpenAI or Ollama - it completely replaces them. You are expected to DISABLE the built-in integrations, and talk with the models implemented by the Pipe function.  Whereas the Pipe function will be doing the heavy-lifting of making the API requests to the respective backend provider.

This discussion thread had 2 or 3 links to the existing Pipe functions that were doing that, for OpenAI provider.

**Reply by GregorBiswanger** [reply]
@DmitriyAlergant-T1A @samling 

Thank you so much for your help! I finally got it. Pipes completely take over the standard connection. I built a function for this that finally works. Thank you!

https://openwebui.com/f/gregorbiswanger/ollama_api_facade_metadata/

If you notice anything in the code that isn't good, then I'd be happy to receive tips. Because I only built it with ChatGPT and am not a Python dev.

Thanks again :)

**Reply by qymab** [reply]
> I'm interested in the solution to this as well. I have open-webui sending to litellm which sends traces to langfuse; I'd like to set tags and other data in langfuse for requests from OWUI, which I'm attempting to do through a filter. I'm doing essentially the same thing as @GregorBiswanger without any luck; looking at the debug logs in litellm, `__metadata__` nor `body["metadata"]` make it to litellm.

Have you found a solution yet?

**Reply by TyZeer** [reply]
@GregorBiswanger 
How do you call this function inside the pipeline or it just gives me all the metadata in body?

**Comment by simonfrey** [comment]
## ✅✅ Fixed. See first comment

I also would need this, especially WITHOUT pipes, as I am currently not working with that abstraction. Idea I had:

**Let's pass the user.id (CREATE TABLE IF NOT EXISTS "user" ("id" VARCHAR(255) NOT NULL) from the database passed as X-Auth-Userid to every upstream/model call.** (Same could be done for other session-related IDs mentionen in the original issue e.g. `id` & `session_id`)

**Security & Privacy concerns**

As the user.id is a random uuid, I consider this safe in terms of user-privacy as well as regarding security, as the user.id is not used for any login purposes.

**Context of my usage**

I am building my own model endpoints, as I want to control what models the users can talk to. e.g. I host my own openai-compatible chat-completion endpoint under localhost:8080/llm/v1/chat/completions.

In order to enable tool usage in my backend, I need a way to identify the single user (e.g. to allow the to only access THEIR OWN google drive files)

**Reply by simonfrey** [reply]
**Seems like there is an undocumenten env var doing exactly what we need 🕺**

`ENABLE_FORWARD_USER_INFO_HEADERS=true` if that is set, it forwards user info:

```python
 if ENABLE_FORWARD_USER_INFO_HEADERS:
            headers["X-OpenWebUI-User-Name"] = user.name
            headers["X-OpenWebUI-User-Id"] = user.id
            headers["X-OpenWebUI-User-Email"] = user.email
            headers["X-OpenWebUI-User-Role"] = user.role
```

src: https://github.com/open-webui/open-webui/blob/main/backend/open_webui/apps/openai/main.py#L594-L598

**Reply by GregorBiswanger** [reply]
@simonfrey  Also what I wanted or am I just missing it?

**Reply by simonfrey** [reply]
Nah, you are missing the session_id  I guess :) This headers identify a single user, but don't allow for building sticky sessions.

**Comment by codysandahl** [comment]
I needed to pass this information along to the Pipelines server, so I wrote a Pipe that packages the extra metadata into the body before sending it along to Pipelines.
https://openwebui.com/f/codysandahl/pipeline_enriched_with_metadata

**Reply by Nj-kol** [reply]
Hi @codysandahl 

This pipe looks good, but how are you able to invoke it? Because I see the in the code https://github.com/open-webui/pipelines/blob/main/main.py#L686

```python
            def stream_content():
                res = pipe(
                    user_message=user_message,
                    model_id=pipeline_id,
                    messages=messages,
                    body=form_data.model_dump(),
                )

```

The `pipe` function is never invoked with metadata from  the pipeline server

**Reply by plamenk** [reply]
Hi @codysandahl, I am new to Open WebUI and have the same question to you like @Nj-kol : what is the trick to get this promissing function running and see the additional info (chat_id, user) within the request body at the backend. Feedback much appreciated!
Plamen

**Reply by codysandahl** [reply]
@plamenk @Nj-kol  I updated the Function to pass along all metadata along with some extra notes for usage.

This Pipe is added inside the OpenWebUI interface as a Function in the Admin Panel. This REPLACES your connection to the Pipelines server if you have it inside the Connections screen of the Admin Panel.

Updated version with some extra notes:
https://openwebui.com/f/codysandahl/pipeline_enriched_with_metadata

**Reply by plamenk** [reply]
Many thanks, @codysandahl . I will test it after I return from the Hanover Fair. For now, I just created two instances, not very smart, I confess, but good enough for a prototype.

**Reply by Nj-kol** [reply]
@codysandahl Thank you so much. Fantastic work!. I can confirm that I can now receive metadata information on the remote pipeline server

**Remote pipeline server code changes**
![Screenshot 2025-03-26 093955](https://github.com/user-attachments/assets/a8567b4c-81f9-44e5-bb2e-bfe4f9bbe9a3)

**Remote pipeline sample output**
![Screenshot 2025-03-26 093854](https://github.com/user-attachments/assets/fa5c5d2b-9a0d-486e-96d1-a759a0f9ca61)

**Reply by lwc943788286-bit** [reply]
@codysandahl Thanks for your great work! I can recieve the metadata of conversation now.
<img width="780" height="620" alt="image" src="https://github.com/user-attachments/assets/20643163-f83b-4e03-b873-69fb76e745c5" />

**Reply by BarrettLowe** [reply]
Hey @codysandahl and @lwc943788286-bit, What version of OWUI are you using? Mine seems to freeze up when I enable the function? I'm on 0.6.34.

**Comment by veleinfobip** [comment]
With function, I need to extract fields from JSON data stored in a database, 
specifically from columns called meta and chat in a table named chat. 
I have try with this:

```  
async def action(
        self,
        body: dict,
        __user__: dict = None,
        __metadata__: dict = None,
        __event_emitter__: Callable[[dict], Any] = None,
        __event_call__: Callable[[dict], Any] = None,
    ) -> None:

        print(f"action:{__name__}")

        metadata = {
            "chat_id": body.pop("chat_id", None),
            "message_id": body.pop("id", None),
            "session_id": body.pop("session_id", None),
            "tool_ids": body.get("tool_ids", None),
            "files": body.get("files", None),
            "title": body.get("title", None),
            "tags": body.get("tags", None),
        }
        body["metadata"] = metadata
        metadata = __metadata__ or body.get("metadata")
```
Output looks like this:
`Metadata: {'chat_id': 'ba28f937-6807-407e-b820-f96ba06aa681', 'message_id': '8418c4aa-654f-4985-ac6f-5f0dbc350780', 'session_id': 'pgjuSEyrSIlJiQMbAABV', 'tool_ids': None, 'files': None, 'title': None, 'tags': None}`

no title or tags..

Thank you in advance for your help!

**Comment by bram49** [comment]
Any updates on this?
With session_id / chat_id, it becomes possible to store states/memory within pipelines.

An easy solution would be to pass chat/session ids within the body dict, upon calling a pipeline.
Just like the user metadata is passed.

**Reply by bram49** [reply]
Think adding metadata to the body would be possible using filters.

Like this Langfuse filter seems to do:
https://github.com/open-webui/pipelines/blob/main/examples/filters/langfuse_filter_pipeline.py 

But would be better if this would be added by default.

Update: This works

**Reply by bram49** [reply]
The body in Filter receives all this:
```
root
├── stream
├── model
├── messages
│   ├── [0]
│       ├── role
│       ├── content
├── features
│   ├── image_generation
│   ├── code_interpreter
│   ├── web_search
├── metadata
│   ├── user_id
│   ├── chat_id
│   ├── message_id
│   ├── session_id
│   ├── tool_ids
│   ├── files
│   ├── features
│   │   ├── image_generation
│   │   ├── code_interpreter
│   │   ├── web_search
│   ├── variables
│   │   ├── {{USER_NAME}}
│   │   ├── {{USER_LOCATION}}
│   │   ├── {{CURRENT_DATETIME}}
│   │   ├── {{CURRENT_DATE}}
│   │   ├── {{CURRENT_TIME}}
│   │   ├── {{CURRENT_WEEKDAY}}
│   │   ├── {{CURRENT_TIMEZONE}}
│   │   ├── {{USER_LANGUAGE}}
│   ├── model
│   │   ├── id
│   │   ├── name
│   │   ├── object
│   │   ├── created
│   │   ├── owned_by
│   │   ├── pipeline
│   │   │   ├── type
│   │   │   ├── valves
│   │   ├── openai
│   │   │   ├── id
│   │   │   ├── name
│   │   │   ├── object
│   │   │   ├── created
│   │   │   ├── owned_by
│   │   │   ├── pipeline
│   │   │       ├── type
│   │   │       ├── valves
│   ├── urlIdx
│   ├── info
│   │   ├── id
│   │   ├── user_id
│   │   ├── base_model_id
│   │   ├── name
│   │   ├── params
│   │   ├── meta
│   │   │   ├── profile_image_url
│   │   │   ├── description
│   │   │   ├── capabilities
│   │   │   │   ├── vision
│   │   │   │   ├── citations
│   │   │   ├── suggestion_prompts
│   │   │   ├── tags
│   │   │   ├── filterIds
│   │   ├── access_control
│   │   │   ├── read
│   │   │   │   ├── group_ids
│   │   │   │   ├── user_ids
│   │   │   ├── write
│   │   │       ├── group_ids
│   │   │       ├── user_ids
│   │   ├── is_active
│   │   ├── updated_at
│   │   ├── created_at
│   ├── actions
│   ├── direct
```

While the body in Pipe, only receives:
```
root
├── stream
├── model
├── messages
│   ├── [0]
│       ├── role
│       ├── content
├── user
│   ├── name
│   ├── id
│   ├── email
│   ├── role
```

**Reply by codysandahl** [reply]
@bram49 I haven't figured out how to use a Filter to pass this metadata because the body gets stripped of extra values. I could be missing something, but that's why I used a Pipe to sit between OpenWebUI and the Pipelines server so I can add the extra values in.

**Reply by codysandahl** [reply]
@bram49 as far as I can tell, OpenWebUI strips out the extra fields before calling the Pipelines server to maintain compatibility with OpenAI's API. If you pass along the extra data directly, it will break an OpenAI API call. So if we add the extra metadata, we'll also have to strip it out before calling an OpenAI API later inside the Pipelines server.

Again, I'm not super confident in this answer. I've only spent a small bit of time on this, so I haven't looked at every angle.

**Reply by bram49** [reply]
@codysandahl Thanks for your reply!
Yesterday I made something similar to your pipeline work with filter (filters already are being called before (inlet) and after (outlet) a pipe, so no need to handle api calls yourself)

A simplified filter to add just the chat_id to the body works as follows:
```python
from typing import Optional

class Filter:
    def __init__(self):
        self.type = "filter"
        self.name = "Add chat_id to body filter"

    async def on_startup(self):
        pass

    async def on_shutdown(self):
        pass

    async def inlet(self, body: dict, user: Optional[dict] = None) -> dict:
        chat_id = body.get("chat_id")
        if chat_id is None:
            chat_id = body.get("metadata") ["chat_id"]       
            
        body["chat_id"] = chat_id

        return body  

    async def outlet(self, body: dict, user: Optional[dict] = None) -> dict:
        """
        Outlet handles the response body (usually the assistant message).
        It will finalize/end the generation created for the user request.
        """
        return body
```

This can also be copied over to the functions settings, and need to be applied in the Models setting for the specific pipeline you want to call.

**Reply by T-Sunm** [reply]
@bram49   Thank you so much. Fantastic work!.

**Comment by jakobhanna** [comment]
Hi, what about the feature to add the additional Metadata to the header of the OWUI backend? Would be very helpful for real load balancing with HAProxy. Could help, but maybe there are othere concerns, whit the `chat_id` is not added as header?

**Comment by ChellappaKannabiran** [comment]
Is this implemented? I would like to pass session-id/chat-id as additional headers/metadata to my custom RAG endpoint.

**Reply by jakobhanna** [reply]
Header ist added with Version v0.6.17 .
